### PR TITLE
using setting api for stat

### PR DESCRIPTION
**Pull request by cyberliem**

### CODE REVIEW

File: stat/fetcher.go [hunk 1]
```diff
@@ -961,25 +977,36 @@ func getTimestampFromTimeZone(t uint64, freq string) uint64 {
 	return result
 }
 
+// getTokenFromAddress do a in-mem lookup for a token matched with the input address
+// if not found, it will attempt to get it from core.
+// if both measure still return no token, stat will panic
+func (self *Fetcher) getTokenFromAddress(addr ethereum.Address) common.Token {
```
Comment (favadi) [inline stat/fetcher.go]: should be renamed to mustGetTokenFromAddress.

File: stat/fetcher.go [hunk 2]
```diff
@@ -961,25 +977,36 @@ func getTimestampFromTimeZone(t uint64, freq string) uint64 {
...
+	var err error
+	token, ok := self.addressLookup[addr]
+	if !ok {
+		token, err = self.setting.GetTokenByAddress(addr)
+		if err != nil {
+			log.Panicf("GetTradeInfo: can't get Token with address %s (err: %s)", addr.Hex(), err)
```
Comment (favadi) [inline stat/fetcher.go]: I think stat should retry few times before panic.

File: http/server.go [hunk 3]
```diff
@@ -1511,6 +1511,13 @@ func (self *HTTPServer) register() {
 	stt.POST("/update-deposit-address", self.UpdateDepositAddress)
 	stt.POST("/update-exchange-info", self.UpdateExchangeInfo)
 	stt.GET("/all-settings", self.GetAllSetting)
+	stt.GET("/internal-tokens", self.GetInternalTokens)
+	stt.GET("/active-tokens", self.GetActiveTokens)
+	stt.GET("/token-by-address", self.GetTokenByAddress)
+	stt.GET("/active-token-by-id", self.GetActiveTokenByID)
+	stt.GET("/address", self.GetAddress)
+	stt.GET("/addresses", self.GetAddresses)
+	stt.GET("/ready-to-serve", self.ReadyToServe)
```
[the remark this answers is not in the document]
Comment (cyberliem) [reply http/server.go]: fix in hash https://github.com/KyberNetwork/reserve-data/pull/397/commits/40057f619d7c9a33ed59d1bf58596d1e0280d797

File: http/settings_handler.go [hunk 4]
```diff
@@ -630,3 +630,105 @@ func (self *HTTPServer) GetAllSetting(c *gin.Context) {
...
+	}
+	httputil.ResponseSuccess(c, httputil.WithData(tokens))
+}
+
+func (self *HTTPServer) GetTokenByAddress(c *gin.Context) {
+	_, ok := self.Authenticated(c, []string{}, []Permission{RebalancePermission, ConfigurePermission, ReadOnlyPermission, ConfirmConfPermission})
+	if !ok {
+		return
+	}
+	addr := c.Query("address")
```
[the remark this answers is not in the document]
Comment (cyberliem) [reply http/settings_handler.go]: fix in hash https://github.com/KyberNetwork/reserve-data/pull/397/commits/40057f619d7c9a33ed59d1bf58596d1e0280d797

File: settings/setting_client.go [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+		Timeout: timeOut,
+	}
+	return &SettingClient{authEngine: authEng,
+		client:  httpClient,
+		coreURL: url}
+}
+
+// SortByKey sort all the params by key in string order
+// This is required for the request to be signed correctly
+func SortByKey(params map[string]string) map[string]string {
```
[the remark this answers is not in the document]
Comment (cyberliem) [reply settings/setting_client.go]: fix in hash https://github.com/KyberNetwork/reserve-data/pull/397/commits/40057f619d7c9a33ed59d1bf58596d1e0280d797

File: settings/setting_client.go [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+	for _, key := range keys {
+		newParams[key] = params[key]
+	}
+	return newParams
+}
+
+func (sc *SettingClient) MakeSign(req *http.Request, message string, nonce string) {
```
[the remark this answers is not in the document]
Comment (cyberliem) [reply settings/setting_client.go]: fix in hash https://github.com/KyberNetwork/reserve-data/pull/397/commits/40057f619d7c9a33ed59d1bf58596d1e0280d797

File: stat/fetcher.go [hunk 7]
```diff
@@ -961,36 +981,52 @@ func getTimestampFromTimeZone(t uint64, freq string) uint64 {
 	return result
 }
 
+// getTokenFromAddress do a in-mem lookup for a token matched with the input address
+// if not found, it will attempt to get it from core.
+// if both measure still return no token, stat will panic
+func (self *Fetcher) mustGetTokenFromAddress(addr ethereum.Address) common.Token {
+	var err error
+	token, ok := self.addressLookup[addr]
```
Comment (favadi) [inline stat/fetcher.go]: I don't see where the resolved name stored in here. Also, it should be protected by a Mutex.

File: stat/fetcher.go [hunk 8]
```diff
@@ -961,36 +981,52 @@ func getTimestampFromTimeZone(t uint64, freq string) uint64 {
...
+	if !ok {
+		for i := 0; i < retryCount; i++ {
+			token, err = self.setting.GetTokenByAddress(addr)
+			if err == nil {
+				break
+			}
+			time.Sleep(retryWait)
+		}
+		if err != nil {
+			log.Panicf("GetTradeInfo: can't get Token with address %s (err: %s)", addr.Hex(), err)
```
Comment (favadi) [inline stat/fetcher.go]: maybe we could just retry until we got the name.

Comment (cyberliem) [reply stat/fetcher.go]: fix in hash https://github.com/KyberNetwork/reserve-data/pull/397/commits/be21a170b57ac878344db6f1a5f3fa18ba154f2c

**Pull request merged**